csrc/cpu/fp8_utils.h

```diff
@@ -0,0 +1,256 @@
+}
+
+static inline __m512i _mm512_cvte5m2_noinf_fp16(__m256i a) {
+  const __m512i vExp_fp16 = _mm512_set1_epi16(0x000F);
+  const __m512i vExp_e5m2 = _mm512_set1_epi16(0x0010);
+  const __m512i vsMant = _mm512_set1_epi16(0x83FF);
+  const __m512i vnaninf = _mm512_set1_epi16(0x8000); /* -0.0 as INF */
+  const __m512i vinfval = _mm512_set1_epi16(0x7c00);
+  __m512i a_ = _mm512_slli_epi16(_mm512_cvtepi8_epi16(a), 8);
+  const __mmask32 mask1_ = _mm512_cmp_epi16_mask(a_, vnaninf, _MM_CMPINT_EQ);
+  __m512i vExp_ = _mm512_sub_epi16(
+      _mm512_srli_epi16(_mm512_and_si512(a_, vinfval), 10), vExp_e5m2);
+  vExp_ = _mm512_slli_epi16(_mm512_add_epi16(vExp_, vExp_fp16), 10);
+  a_ = _mm512_or_si512(vExp_, _mm512_and_si512(a_, vsMant));
+  return _mm512_mask_mov_epi16(a_, mask1_, vinfval);
+}
+
+static inline void cvt_fp16_e5m2_noINF_rne_intrinsic(
+    const short* __restrict__ in, unsigned char* out, int size) {
+#pragma omp parallel for
+  for (int i = 0; i < size; i += 32) {
+    __m256i bh_ = _mm256_lddqu_si256((__m256i*)&in[i]);
+    __m256i ah_ = _mm256_lddqu_si256((__m256i*)&in[i + 16]);
+    _mm256_storeu_si256((__m256i*)&out[i], _mm256_cvt2fp16_e5m2(ah_, bh_));
+  }
+}
+
+static inline void cvt_fp32_e5m2_noinf_rne_intrinsic(
+    const float* __restrict__ in, float* out, int size, float scale) {
+#pragma omp parallel for
+  for (int i = 0; i < size; i += 32) {
+    __m512 b = _mm512_loadu_ps(&in[i]);
+    __m512 a = _mm512_loadu_ps(&in[i + 16]);
+    __m256i ah_ =
+        _mm512_cvtps_ph(a, (_MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC));
+    __m256i bh_ =
+        _mm512_cvtps_ph(b, (_MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC));
+    __m512i a_rne_ =
+        _mm512_cvte5m2_noinf_fp16(_mm256_cvt2fp16_e5m2_noINF(ah_, bh_));
+    bh_ = _mm512_extracti64x4_epi64(a_rne_, 0);
+    ah_ = _mm512_extracti64x4_epi64(a_rne_, 1);
+    b = _mm512_cvtph_ps(bh_);
+    a = _mm512_cvtph_ps(ah_);
+    _mm512_storeu_ps(&out[i], b);
+    _mm512_storeu_ps(&out[i + 16], a);
+  }
+}
+
+static inline __m512i cast_fp8x32_to_fp16x32(__m256i a) {
+  return _mm512_cvte5m2_fp16(a);
+}
+
+static inline __m256i cast_fp16x16x2_to_fp8x32(__m256i a, __m256i b) {
+  return _mm256_cvt2fp16_e5m2(a, b);
+}
+
+static inline void cast_fp8xn_to_fp16xn(const char* __restrict__ in,
+                                        unsigned short* out, int n) {
+#pragma omp parallel for
+  for (int i = 0; i < n; i += 32) {
+    __m256i a = _mm256_loadu_si256((const __m256i*)&in[i]);
+    __m512i b = cast_fp8x32_to_fp16x32(a);
+    __m512i* out_p = (__m512i*)(&out[i]);
+    *out_p = b;
+  }
+}
+
+static inline void cast_fp16xn_to_fp8xn(const short* __restrict__ in,
+                                        unsigned char* out, int n) {
+  cvt_fp16_e5m2_noINF_rne_intrinsic(in, out, n);
+}
+
+static inline void cast_fp32xn_to_fp8xn(const float* __restrict__ in,
+                                        float* out, int n) {
+  cvt_fp32_e5m2_noinf_rne_intrinsic(in, out, n, 0);
+}
```

Comment:
It seems you aren't getting any reuse out of these functions, why not inline the body?